src/box/relay.cc

```diff
@@ -637,12 +638,25 @@ tx_status_update(struct cmsg *msg)
 
 /**
  * Update replica gc state in tx thread.
+ * Note that there is a case when replica and its consumer were dropped
+ * from system spaces, but the relay is not stopped yet - we must not create
+ * new gc consumer in this case, because it will become hanging. So the
+ * function tries to update gc_consumer - it is not created if it doesn't
+ * exist.
  */
 static void
 tx_gc_advance(struct cmsg *msg)
 {
 	struct relay_gc_msg *m = (struct relay_gc_msg *)msg;
-	gc_consumer_advance(m->relay->replica->gc, &m->vclock);
+	struct credentials *old_user = effective_user();
+
+	fiber_set_user(fiber(), &admin_credentials);
+	if (box_gc_consumer_update(m->relay->replica, &m->vclock) != 0) {
```

Comment:
@Gerold103 was strongly against creating GC consumers during `SUBSCRIBE` or `FETCH_SNAPSHOT`. However, this allows us to drop a row from `_gc_consumers`, when user knows, that replica will be down for a long time without deleting it from `_cluster`, as replica will create a new GC consumer during resubscribe. It also simplifies the join protocol for anon replicas.

If we prohibit creating GC consumer during `SUBSCRIBE` and make only update of `_gc_consumers` space, then replica will have to make rejoin even if xlogs are not yet deleted, which is not great. So, we should prohibit deletion of a row, while there's a row in `_cluster` in order to always update it during `SUBSRIBE`

If we allow creating GC consumer on `SUBSCRIBE`, then we should also allow doing this on `FETCH_SNAPSHOT` in order to simplify the protocol, `REGISTER` is not needed. If we decide to violate `SUBSCRIBE`, then we should also violate `FETCH_SNAPSHOT`.

@Gerold103 @sergepetrenko @drewdzzz, we have to decide, whether we allow creating new rows during `SUBSCRIBE` and `FETCH_SNAPSHOT`. I'm voting for allowing doing this, as it simplifies the protocol of anon join a lot, and allows to drop consumers without dropping row from `_cluster

Reply:
@Gerold103, TBH I don't remember the motivation behind forbidding to create gc consumers in `SUBSCRIBE` or `FETCH_SNAPSHOT`. The space `_gc_consumers` is local, I don't see anything wrong with writing into it during these requests processing.

Qould you please repeat your arguments here?

Reply:
@Gerold103, thanks for the detailed answer!

> Why do we 1) allow to drop GC for registered replicas at all?

For the sake of easy administration. I'm citing Vlad Grubov here:
From time to time it is known ahead of time that some node will go down due to maintenance. And server maintenance (disk, RAM replacement) may last several days. In this case we want to stop holding xlogs for this replica and simply rejoin it back to cluster once the maintenance is done. Because syncing all the xlogs piled up during this time will take too long. Besides, master may simply run out of disk space due to this stray replica (and usually there are a lot of masters/nodes on the same physical server, so it'll affect a whole bunch of shards).

In this particular scenario there is no difference between dropping a replica from `_gc_consumers` and dropping it from `_cluster`.

Although there is another case: what if the node could actually rejoin just fine? Maybe it is returned soon enough and xlogs for it are still in place. In this case if we delete it from `_cluster` it still has to rejoin. And deleting it only from `_gc_consumers` would let it re-subscribe without joining once again.

Reply:
> I think I am missing some reasoning here.

Good point, I should have started with my motivation. I have mostly technical reasons here. Implementing persistent WAL GC for non-anonymous replica, I faced a choice: should I create gc consumer on `JOIN` or on `SUBSCRIBE`? My thoughts:

## Creating gc consumer for non-anonymous replica on JOIN

Since a replica joins to one instance and then subscribes to the whole replicaset, I should create its consumer on insert to space `_cluster` to create the consumers on all replicas (I need it because after leader's fault, replica's won't rejoin to the new one, so it must have gc consumers before becoming a leader). Here I can create consumer with empty vclock (it will be inactive then) or with current gc vclock (to prevent any cleanup).

The second solution sounds bad: despite recommended Tarantool topology is full-mesh, people can use others as well, for example, a star topology - in this case replicas are not connected with each other, but they still have their rows in `_cluster`, so all the replicas will have active gc consumers for other ones, and they will never be advanced, preventing xlogs from being cleaned up until the first ENOSPC.

So the only good solution is create a consumer with empty vclock on insert in space `_cluster`. But it requires a crutch here - `_gc_consumers` mustn't be created on local recovery (because then we already created it), but it must created on recovery from master and after recovery. Doesn't look good, as for me.

Also, in star topology all replicas will have empty consumers for others for no reason (they will never be used), but it's not critical at all, just a minor addition :)

## Creating gc consumer for non-anonymous replica on SUBSCRIBE

Here, if replica is not anonymous, it passed JOIN phase, so it has a row in `_cluster`. And it has a `_gc_consumer`, but only on bootstrap master. After subscribing to the replicas, the consumer will be created on them as well (and only on the replicas that it's connected to). Here I don't need to create consumers on replace in `_cluster`, and it makes the code easier.

Another advantage of the approach: AFAIK, there are cases when admin manually adds a replica to `_cluster`, manually copies actual snapshot to the replica and simply calls `SUBSCRIBE`. And this approach will work exactly in the same way if we will create consumers on `SUBSCRIBE`, but it will require him to figure out the problem and manually create a consumer if we create them only on `JOIN`. We will add an opportunity to hold xlogs manually with creating gc consumer but we won't force user to use new functionality, and backward compatibility of such complicated case won't be broken.

> You can think of it this way - this instance is struct instance { struct gc_consumer **consumers; }. Inside it has an array of struct gc_consumer * pointers. Changing the consumers[i]->... members won't change the instance itself. But inserting a new consumer into consumers array means updating instance itself. Its own direct member.

I see it in the same way, but with one difference. If replica has a row in `_cluster`, but has no row in `_gc_consumer`, it has a consumer in this array of consumers, but it's empty. In these terms, `SUBSCRIBE` is still read-only and does not change state of the instance. My abstraction is a bit more complicated, but I prefer it more than workaround with creating consumer only on remote recovery and after recovery.


## Addressing some comments


> Update of _gc_consumers with a new vclock might be considered not a state change, since both before and after that the state was "I am holding the GC on for this replica".

Actually, after join there will be an empty inactive gc consumer on all replicas except for bootstrap master (see `JOIN` section for rationale), and only after subscribe the consumer will become active and will prevent xlogs from being cleaned up, so both approaches are the same on this matter.

> Consider a similar example - if a user is granted RW rights to a certain space, it doesn't mean they must be able to write into that space using space:select(), right?

Yep, good example. But when using online space upgrade, `space:select` can create a new tuple (with applied upgrade function) - this lazy creation of tuple, as for me, is the same as lazy creation of `gc_consumer` for registered replica.

Reply:
Discussed f2f, decided to create dummy consumer on `_cluster:insert` and forbid to create it on `SUBSCRIBE`.

Reply:
@Serpentian Here's what we've come up with during a verbal discussion with @Gerold103 and @drewdzzz.

When replica registers (joins) on master, master inserts row in `_cluster`.
On-replace trigger in `_cluster` creates a new row for the replica in `_gc_consumers`. Delete from `_cluster` also deletes from `_gc_consumers`. This row is "dummy". It doesn't contain a vclock.
Anon replicas must use REGISTER to create a dummy entry for them in `_gc_consumers` and then SUBSCRIBE. To change source instance the anon replica must resend REGISTER to the new instance it is going to SUBSCRIBE to.
If an anon replica tries to SUBSCRIBE and there is no `_gc_consumers` entry for it, SUBSCRIBE fails.
(for anon replicas) delete from `_gc_consumers` on timeout if no entry in `_cluster`.

If a user wants to stop holding xlogs for a replica, he replaces its gc consumer with a dummy (with no vclock). Deleting a _gc_consumers entry for a node which is present in _cluster is prohibited.

Later on we may design some interface to manipulate replica registrations: drop gc conumer, add a replica with a dummy gc consumer and so on via different `box.ctl.*()` calls. To make it easier than just replacing system spaces for the user.